### PR TITLE
Return BadRequestException with valid message when input json body is invalid

### PR DESCRIPTION
**Pull request by apurvabhaleMS**
## Description
For invalid json body requests Fhir server returns 500. If the input json is not in the correct format then we are parsing the body at [FhirJsonInputFormatter](https://github.com/microsoft/fhir-server/blob/main/src/Microsoft.Health.Fhir.Shared.Api/Features/Formatters/FhirJsonInputFormatter.cs) and passing the initial validations for required fields at [ModelAttributeValidator](https://github.com/microsoft/fhir-server/blob/main/src/Microsoft.Health.Fhir.Core/Features/Validation/IModelAttributeValidator.cs).

 e.g. 
If the body contains _Coverage.status = ""_, then after parsing _Coverage.status = null_ & _Coverage.statusElement = null_, resulting into minimum cardinality error as expected.
If the body contains _Coverage.status =_ , then after parsing _Coverage.status = null_ & _Coverage.statusElement = {value=null}_, which passes the Firely validation and [CodeToTokenSearchValueConverter](https://github.com/microsoft/fhir-server/blob/main/src/Microsoft.Health.Fhir.Core/Features/Search/Converters/CodeToTokenSearchValueConverter.cs) returns null causing search index entry to fail.

 In this PR we are returning BadRequestException with a valid message instead of 500

## Related issues
Addresses [85266](https://microsofthealth.visualstudio.com/Health/_workitems/edit/85266)

## Testing
1. Check POST Coverage request with - "status" : ,
2. Check POST Member-Match request with - "status" : ,
3. Check POST Claim request with - "use": , 
4. Unit tests

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [x] Tag the PR with Azure API for FHIR if this will release to the Azure API for FHIR managed service (CosmosDB or common code related to service)
- [x] Tag the PR with Azure Healthcare APIs if this will release to the Azure Healthcare APIs managed service (Sql server or common code related to service)
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/TypedElementSearchIndexerTests.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+using Microsoft.Health.Fhir.Core.Features.Search;
+using Microsoft.Health.Fhir.Core.Features.Search.Converters;
+using Microsoft.Health.Fhir.Core.Models;
+using Microsoft.Health.Fhir.Tests.Common;
+using NSubstitute;
+using Xunit;
+
+namespace Microsoft.Health.Fhir.Core.UnitTests.Features.Search
+{
+    public class TypedElementSearchIndexerTests
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/TypedElementSearchIndexerTests.cs:25]: There was a lot of work put into getting the search indexer out of "Shared" and into FHIR agnostic core, a little sad to see it back :(

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchIndexEntry.cs [hunk 2]
```diff
@@ -18,12 +18,26 @@ public class SearchIndexEntry : IEquatable<SearchIndexEntry>
...
-            EnsureArg.IsNotNull(value, nameof(value));
+            if (searchParameter is null)
+            {
+                throw new BadRequestException(string.Format(Resources.ValueCannotBeNull, "Search Parameter"));
+            }
+
+            if (value is null)
+            {
+                throw new BadRequestException(string.Format(Resources.ValueCannotBeNull, searchParameter.Expression));
+            }
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Search/SearchIndexEntry.cs:40]: Should the search indexer handle this instead? This still looks like an ArgumentException in this class.

```
  foreach (ISearchValue searchValue in ExtractSearchValues(
      searchParameter.Url.ToString(),
      searchParameter.Type,
      searchParameter.TargetResourceTypes,
      resource,
      searchParameter.Expression,
      context))
  {
     if(searchValue != null) {
      yield return new SearchIndexEntry(searchParameterInfo, searchValue);
     }
  }
```

or

```
  foreach (ISearchValue searchValue in ExtractSearchValues(
      searchParameter.Url.ToString(),
      searchParameter.Type,
      searchParameter.TargetResourceTypes,
      resource,
      searchParameter.Expression,
      context))
  {
     try {
      yield return new SearchIndexEntry(searchParameterInfo, searchValue);
     } catch(ArgumentException) { throw new BadRequestException(...) }
  }
```

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/TypedElementSearchIndexerTests.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            FhirTypedElementToSearchValueConverterManager fhirTypedElementToSearchValueConverterManager = await SearchParameterFixtureData.GetFhirTypedElementToSearchValueConverterManagerAsync();
+            return fhirTypedElementToSearchValueConverterManager;
+        }
+
+        [Fact]
+        public void GivenAValidResource_WhenExtract_ThenValidSearchIndexEntriesAreCreated()
+        {
+            var coverageResource = Samples.GetDefaultCoverage().ToPoco<Coverage>();
+
+            var test = _searchIndexer.Extract(coverageResource.ToResourceElement());
```
[the remark this answers is not in the document]
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/TypedElementSearchIndexerTests.cs:68]: These also look like they should be collapsed:

```
 var coverageResource = Samples.GetDefaultCoverage();
 var test = _searchIndexer.Extract(coverageResource);
```

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/TypedElementSearchIndexerTests.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+#if !Stu3
+        // For Stu3 - Coverage.status, Observation.status, and Claim.use are not required fields
+        [Fact]
+        public void GivenAnInValidResource_WhenExtract_ThenExceptionIsThrown()
+        {
+            var requestBundle = Samples.GetJsonSample("Bundle-TransactionWithInvalidBundleEntry").ToPoco<Bundle>();
+
+            foreach (var entry in requestBundle.Entry)
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/TypedElementSearchIndexerTests.cs:78]: Something like:

```suggestion
            var requestBundle = Samples.GetJsonSample("Bundle-TransactionWithInvalidBundleEntry");

            foreach (var entry in new BundleEntryWrapper(requestBundle.Instance).Entries)
```

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/TypedElementSearchIndexerTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (apurvabhaleMS) [reply src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/TypedElementSearchIndexerTests.cs]: Oops missed the Assert part. Updated the tests with the latest commit

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchIndexEntry.cs (hunk 2 above)
Comment (apurvabhaleMS) [reply src/Microsoft.Health.Fhir.Core/Features/Search/SearchIndexEntry.cs]: Added the null check in TypedElementSearchIndexer.ProcessNonCompositeSearchParameter. I noticed there is another function ProcessCompositeSearchParameter which calls the SearchIndexEntry too. Wondering if the null check would also apply to the composite Search parameter. Any thoughts?

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/TypedElementSearchIndexerTests.cs (hunk 4 above)
Comment (apurvabhaleMS) [reply src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/TypedElementSearchIndexerTests.cs]: Updated the test as per suggestion.

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/TypedElementSearchIndexerTests.cs (hunk 1 above)
Comment (apurvabhaleMS) [reply src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Search/TypedElementSearchIndexerTests.cs:27]: :( I can see if I can move this test to Microsoft.Health.Fhir.Core.UnitTests from Microsoft.Health.Fhir.Shared.Core.UnitTests

**Review by LTA-Thinking — APPROVED** [review]

**Comment by brendankowitz** [comment]
Possibly related to AB#86850

**Pull request merged**